### ISSUE TITLE
Better benchmarks

### ISSUE THREAD

**Issue by steveklabnik** [steveklabnik/indexlist#1]
The benchmarks are... tough. I'm not great with criteron, so I have to massively pre-allocate to test pop_front: https://github.com/steveklabnik/indexlist/blob/master/benches/benchmarks.rs#L99

There's gotta be a better way.

**Comment by daniel-ferguson** [comment]
I couldn't run the benchmarks on my machine due to the pre-allocation you mentioned, the process got OOM'd after trying to allocate more than my 16GB of RAM (I'm not sure this is expected, nonetheless it motivated me to trying to find a better way). 

I had a quick go at re-implementing them differently and arrived at the following - worth noting that this is the first time I've used criterion so there might be much better approaches.

```rust
c.bench_function_over_inputs(
    "pop_front_index_list",
    |b: &mut criterion::Bencher, size: &usize| {
        b.iter_with_large_setup(
            || {
                let mut list = IndexList::new();
                for i in 0..*size {
                    list.push_back(i);
                }
                (list, *size)
            },
            |(mut list, len)| {
                for _ in 0..len {
                    list.pop_front().unwrap();
                }
            },
        );
    },
    vec![8, 64, 1024, 2048, 4096],
);

c.bench_function_over_inputs(
    "pop_front_linked_list",
    |b: &mut criterion::Bencher, size: &usize| {
        b.iter_with_large_setup(
            || {
                let mut list = LinkedList::new();
                for i in 0..*size {
                    list.push_back(i);
                }
                (list, *size)
            },
            |(mut list, len)| {
                for _ in 0..len {
                    list.pop_front().unwrap();
                }
            },
        );
    },
    vec![8, 64, 1024, 2048, 4096],
);
```

Questions:
* I'm not sure `iter_with_large_setup` is necessary, perhaps `iter_with_setup` is fine in this case
* This changes the thing being measured from "how long does a `pop_front()` take?" to "given a list of length `n`, how long does it take to clear it using `pop_front()`?" - I'm not sure this is as meaningful, or easy to interpret
* Is the range of list sizes useful? On my machine it seems just broad enough to show how `LinkedList` and `IndexList` diverge in performance as input size grows

**Comment by daniel-ferguson** [comment]
I think I've stumbled upon a better approach, it:
* produces the helpful function comparison pages your current implementation does
* covers a range of list sizes, and presents their performance characteristics on the same page in the criterion report
* is much closer to the approach used in the other benchmarks
* feels like a *little* bit of a hack
* benchmarks lists of different sizes (like the approach above)
* could be trivially adapted to only `pop_front()` once per iteration - as your current benchmark does

```rust
let mut functions = vec![];

for list_length in &[1, 1_000, 10_000, 10_000, 100_000] {
    let list_length = *list_length;

    let list = Fun::new(&format!("linked_list{}", list_length), move |b, _| {
        b.iter_with_setup(
            || {
                let mut list = LinkedList::new();
                for i in 0..list_length {
                    list.push_back(i);
                }
                list
            },
            |mut list| {
                for _ in 0..list_length {
                    list.pop_front().unwrap();
                }
            },
        );
    });

    // same for index_list

    functions.push(list);
    functions.push(index_list);
}

// no input
c.bench_functions("pop_front", functions, 0);
```

It feels a bit weird to not be using the input parameter for `bench_functions` but I couldn't find a good way to get that to work with a list of inputs. I think if criterion had a `bench_functions_over_inputs`, that might be the thing you'd want to reach for.

**Comment by steveklabnik** [comment]
I think that sounds great!

> It feels a bit weird to not be using the input parameter for bench_functions but I couldn't find a good way to get that to work with a list of inputs.

Yeah, that was my issue too. It feels like it should be used for this, but it's not clear how. Maybe @japaric has some advice?

**Comment by daniel-ferguson** [comment]
If you'd appreciate a PR let me know - I can understand waiting on input from others with more benchmarking and/or criterion experience. None of the "solutions" I proposed feel quite right, but the last one seems the most ok.

If a PR would be appreciated I think I'd be keen to figure out an answer to the following:
- [ ] what are we benchmarking, a single `pop_front()`, or emptying a list using `pop_front()`?
- [ ] do we want to benchmark against a single list size, or a range of sizes as illustrated in the last code snippet I posted?
- [ ] if the latter, does a range between 1 and 100,000(?) in powers of 10 increments seem like a reasonable start, or a bit over the top? 

    *Currently formatting those into the test names results in names which are quite hard to distinguish on reports: differentiating between 1000 and 10000 in a proportional font is a little awkward for me. Possibly a case for powers of 2 numbers.*
- [ ] are we interested in making changes to the other benchmarks to also operate over a range of list sizes?

  *This could well be out of scope and a question for another time/issue, but if introducing inconsistency feels irksome, it might be straightforward to tackle it as part of the same PR.*
